dvc/utils/stream.py

```diff
@@ -43,3 +46,17 @@ def read1(self, n=-1):
 
         output, self.leftover = chunk[:n], chunk[n:]
         return output
+
+    read = read1
+
+    def peek(self, n: int):
+        if self.leftover and len(self.leftover) >= n:
+            return self.leftover[:n]
+        data = b""
+        while len(data) < n:
+            chunk = self.read1(n - len(data))
+            if not chunk:
+                break
+            data += chunk
+        self.leftover = data + self.leftover if self.leftover else data
```

Comment:
Still can't do that. I.e. the whole content is `123456789`. We make `.read(2)` it returns `12` and `34` goes to leftover. Than you `.peek(7)` and get `5678934`, which is wrong.

Reply:
Don't see how you are getting there. 
```
t = IterStream(iter([b"123456789"]))
print(t.read(2))
print(t.peek(7))
print(t.read())
```

Returned
```
b'12'
b'3456789'
b'3456789'
```

Reply:
And 
```
t = IterStream(iter([b"1234", b"56789"]))
print(t.read(2))
print(t.peek(7))
print(t.read())
```
Returned the same

Reply:
Ok, I guess I was wrong here. Probably just looks really counter-intuitive to me)

Reply:
```python
    def __init__(self, iterator):
        ...
        self.leftover = b""  # And set it to b"" everywhere instead of None

    def peek(self, n):
        while len(self.leftover) < n:
            try:
                self.leftover += next(self.iterator)
            except StopIteration:
                break 
        return self.leftover
```
Looks more natural.

Reply:
> ```python
>     def __init__(self, iterator):
>         ...
>         self.leftover = b""  # And set it to b"" everywhere instead of None
> 
>     def peek(self, n):
>         while len(self.leftover) < n:
>             try:
>                 self.leftover += next(self.iterator)
>             except StopIteration:
>                 break 
>         return self.leftover
> ```
> 
> Looks more natural.

Done, take a look.